File: app/src/main/java/br/com/jonathanzanella/myexpenses/expense/EditExpenseActivity.java
package br.com.jonathanzanella.myexpenses.expense;

import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.CheckBox;
import android.widget.EditText;

import org.apache.commons.lang3.StringUtils;
import org.joda.time.DateTime;

import br.com.jonathanzanella.myexpenses.R;
import br.com.jonathanzanella.myexpenses.bill.Bill;
import br.com.jonathanzanella.myexpenses.bill.BillRepository;
import br.com.jonathanzanella.myexpenses.bill.ListBillActivity;
import br.com.jonathanzanella.myexpenses.chargeable.Chargeable;
import br.com.jonathanzanella.myexpenses.chargeable.ChargeableType;
import br.com.jonathanzanella.myexpenses.chargeable.ListChargeableActivity;
import br.com.jonathanzanella.myexpenses.database.RepositoryImpl;
import br.com.jonathanzanella.myexpenses.helpers.CurrencyHelper;
import br.com.jonathanzanella.myexpenses.helpers.CurrencyTextWatch;
import br.com.jonathanzanella.myexpenses.log.Log;
import br.com.jonathanzanella.myexpenses.validations.ValidationError;
import br.com.jonathanzanella.myexpenses.views.BaseActivity;
import butterknife.Bind;
import butterknife.OnClick;

public class EditExpenseActivity extends BaseActivity implements ExpenseContract.EditView {
	public static final String KEY_EXPENSE_UUID = ExpensePresenter.KEY_EXPENSE_UUID;
	private static final int REQUEST_SELECT_CHARGEABLE = 1003;
	private static final int REQUEST_SELECT_BILL = 1004;

	@Bind(R.id.act_edit_expense_name)
	EditText editName;
	@Bind(R.id.act_edit_expense_date)
	EditText editDate;
	@Bind(R.id.act_edit_expense_value)
	EditText editValue;
	@Bind(R.id.act_edit_expense_value_to_show_in_overview)
	EditText editValueToShowInOverview;
	@Bind(R.id.act_edit_expense_repayment)
	CheckBox checkRepayment;
	@Bind(R.id.act_edit_expense_chargeable)
	EditText editChargeable;
	@Bind(R.id.act_edit_expense_bill)
	EditText editBill;
	@Bind(R.id.act_edit_expense_pay_next_month)
	CheckBox checkPayNextMonth;
	@Bind(R.id.act_edit_expense_show_in_overview)
	CheckBox showInOverview;
	@Bind(R.id.act_edit_expense_show_in_resume)
	CheckBox showInResume;
	@Bind(R.id.act_edit_expense_repetition)
	EditText editRepetition;
	@Bind(R.id.act_edit_expense_installment)
	EditText editInstallment;

	private final ExpensePresenter presenter;

	public EditExpenseActivity() {
		ExpenseRepository expenseRepository = new ExpenseRepository(new RepositoryImpl<Expense>(this));
		presenter = new ExpensePresenter(expenseRepository, new BillRepository(new RepositoryImpl<Bill>(this), expenseRepository));
	}

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_edit_expense);
	}

	@Override
	protected void onPostCreate(Bundle savedInstanceState) {
		super.onPostCreate(savedInstanceState);

		editValue.addTextChangedListener(new CurrencyTextWatch(editValue));
		editValueToShowInOverview.addTextChangedListener(new CurrencyTextWatch(editValueToShowInOverview));
		editValue.setOnFocusChangeListener(new View.OnFocusChangeListener() {
			@Override
			public void onFocusChange(View v, boolean hasFocus) {
				if(editValueToShowInOverview.getText().toString().isEmpty() &&
						!editValue.getText().toString().isEmpty()) {
					editValueToShowInOverview.setText(editValue.getText());
				}
			}
		});
		presenter.attachView(this);
		presenter.onViewUpdated(false);
	}

	@Override
	protected void storeBundle(Bundle extras) {
		super.storeBundle(extras);
		if(extras != null)
			presenter.storeBundle(extras);
	}

	@Override
	protected void onSaveInstanceState(Bundle outState) {
		super.onSaveInstanceState(outState);
		presenter.onSaveInstanceState(outState);
	}

	@Override
	protected void onStart() {
		super.onStart();
		presenter.attachView(this);
	}

	@Override
	protected void onResume() {
		super.onResume();
		presenter.refreshExpense();
	}

	@Override
	protected void onStop() {
		super.onStop();
		presenter.detachView();
	}

	@Override
	protected void onActivityResult(int requestCode, int resultCode, Intent data) {
		super.onActivityResult(requestCode, resultCode, data);
		presenter.attachView(this);
		switch (requestCode) {
			case REQUEST_SELECT_CHARGEABLE: {
				if(resultCode == RESULT_OK) {
					String uuid = data.getStringExtra(ListChargeableActivity.Companion.getKEY_CHARGEABLE_SELECTED_UUID());
					String keyChargeableSelectedType = ListChargeableActivity.Companion.getKEY_CHARGEABLE_SELECTED_TYPE();
					ChargeableType type = (ChargeableType) data.getSerializableExtra(keyChargeableSelectedType);
					presenter.onChargeableSelected(type,uuid);
				}
				break;
			}
			case REQUEST_SELECT_BILL: {
				if(resultCode == RESULT_OK)
					presenter.onBillSelected(data.getStringExtra(ListBillActivity.Companion.getKEY_BILL_SELECTED_UUID()));
				break;
			}
		}
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		getMenuInflater().inflate(R.menu.save, menu);
		return true;
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		switch (item.getItemId()) {
			case R.id.action_save:
				presenter.save();
				break;
		}
		return super.onOptionsItemSelected(item);
	}

	@OnClick(R.id.act_edit_expense_date)
	void onBalanceDate() {
		presenter.onDate(this);
	}

	@Override
	public void onDateChanged(DateTime date) {
		editDate.setText(Expense.SIMPLE_DATE_FORMAT.format(date.toDate()));
	}

	@Override
	public void onChargeableSelected(Chargeable chargeable) {
		editChargeable.setText(chargeable.getName());
		checkPayNextMonth.setVisibility(chargeable.canBePaidNextMonth() ? View.VISIBLE : View.GONE);
	}

	@OnClick(R.id.act_edit_expense_chargeable)
	void onChargeable() {
		new AsyncTask<Void, Void, Boolean>() {

			@Override
			protected Boolean doInBackground(Void... voids) {
				return presenter.hasChargeable();
			}

			@Override
			protected void onPostExecute(Boolean hasChargeable) {
				super.onPostExecute(hasChargeable);
				if(!hasChargeable) {
					Intent intent = new Intent(EditExpenseActivity.this, ListChargeableActivity.class);
					startActivityForResult(intent, REQUEST_SELECT_CHARGEABLE);
				}
			}
		}.execute();
	}

	@Override
	public void onBillSelected(Bill bill) {
		if(bill != null) {
			if(editBill.getText().toString().isEmpty())
				editBill.setText(bill.getName());
			if(editName.getText().toString().isEmpty())
				editName.setText(bill.getName());
			if(editValue.getText().toString().isEmpty())
				editValue.setText(CurrencyHelper.format(bill.getAmount()));
			showInOverview.setChecked(false);
			showInResume.setChecked(true);
		} else {
			editBill.setText("");
		}
	}

	@OnClick(R.id.act_edit_expense_bill)
	void onBill() {
		startActivityForResult(new Intent(this, ListBillActivity.class), REQUEST_SELECT_BILL);
	}

	@Override
	public Expense fillExpense(Expense expense) {
		expense.setName(editName.getText().toString());
		int value = 0;
		String valueText = editValue.getText().toString().replaceAll("[^\\d]", "");
		if(!StringUtils.isEmpty(valueText))
			value = Integer.parseInt(valueText);

		int valueToShowInOverview = 0;
		String valueToShowInOverviewText = editValueToShowInOverview.getText().toString().replaceAll("[^\\d]", "");
		if(!StringUtils.isEmpty(valueToShowInOverviewText))
			valueToShowInOverview = Integer.parseInt(valueToShowInOverviewText);

		if (checkRepayment.isChecked()) {
			value *= -1;
			valueToShowInOverview *= -1;
		}
		expense.setValue(value);
		expense.setValueToShowInOverview(valueToShowInOverview);
		expense.setChargedNextMonth(checkPayNextMonth.isChecked());
		expense.showInOverview(showInOverview.isChecked());
		expense.showInResume(showInResume.isChecked());
		expense.setInstallments(getInstallment());
		expense.setRepetition(getRepetition());
		return expense;
	}

	@Override
	public void finishView() {
		Intent i = new Intent();
		i.putExtra(KEY_EXPENSE_UUID, presenter.getUuid());
		setResult(RESULT_OK, i);
		finish();
	}

	@Override
	public void showError(ValidationError error) {
		switch (error) {
			case NAME:
				editName.setError(getString(error.getMessage()));
				break;
			case AMOUNT:
				editValue.setError(getString(error.getMessage()));
				break;
			case CHARGEABLE:
				editChargeable.setError(getString(error.getMessage()));
				break;
			default:
				Log.error(this.getClass().getName(), "Validation unrecognized, field:" + error);
		}
	}

	public int getInstallment() {
		return Integer.parseInt(editInstallment.getText().toString());
	}

	public int getRepetition() {
		return Integer.parseInt(editRepetition.getText().toString());
	}

	@Override
	public void showExpense(Expense expense) {
		editName.setText(expense.getName());
		editValue.setText(CurrencyHelper.format(Math.abs(expense.getValue())));
		editValueToShowInOverview.setText(CurrencyHelper.format(Math.abs(expense.getValueToShowInOverview())));
		if(expense.isCharged()) {
			//noinspection deprecation
			editValue.setTextColor(getResources().getColor(R.color.value_unpaid));
			checkRepayment.setEnabled(false);
		}
		if(expense.getValue() < 0)
			checkRepayment.setChecked(true);

		checkPayNextMonth.setChecked(expense.isChargedNextMonth());
		showInOverview.setChecked(expense.isShowInOverview());
		showInResume.setChecked(expense.isShowInResume());
	}
}

File: app/src/main/kotlin/br/com/jonathanzanella/myexpenses/expense/EditExpenseActivity.kt
package br.com.jonathanzanella.myexpenses.expense

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v4.content.res.ResourcesCompat
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.AppCompatEditText
import android.text.InputType
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.CheckBox
import android.widget.LinearLayout
import br.com.jonathanzanella.myexpenses.R
import br.com.jonathanzanella.myexpenses.bill.Bill
import br.com.jonathanzanella.myexpenses.bill.BillRepository
import br.com.jonathanzanella.myexpenses.bill.ListBillActivity
import br.com.jonathanzanella.myexpenses.chargeable.Chargeable
import br.com.jonathanzanella.myexpenses.chargeable.ChargeableType
import br.com.jonathanzanella.myexpenses.chargeable.ListChargeableActivity
import br.com.jonathanzanella.myexpenses.helpers.CurrencyTextWatch
import br.com.jonathanzanella.myexpenses.helpers.toCurrencyFormatted
import br.com.jonathanzanella.myexpenses.transaction.Transaction
import br.com.jonathanzanella.myexpenses.validations.ValidationError
import br.com.jonathanzanella.myexpenses.views.anko.*
import org.apache.commons.lang3.StringUtils
import org.jetbrains.anko.*
import org.joda.time.DateTime
import timber.log.Timber

class EditExpenseActivity : AppCompatActivity(), ExpenseContract.EditView {
    override val context = this
    private val ui = EditExpenseActivityUi()
    private val presenter: ExpensePresenter

    init {
        val expenseRepository = ExpenseRepository()
        presenter = ExpensePresenter(expenseRepository, BillRepository(expenseRepository))
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ui.setContentView(this)

        storeBundle(savedInstanceState)
        storeBundle(intent.extras)

        ui.toolbar.setup(this)
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)

        ui.value.addTextChangedListener(CurrencyTextWatch(ui.value))
        ui.valueToShowInOverview.addTextChangedListener(CurrencyTextWatch(ui.valueToShowInOverview))
        ui.value.onFocusChangeListener = View.OnFocusChangeListener { _, _ ->
            if (ui.valueToShowInOverview.text.toString().isEmpty() && !ui.value.text.toString().isEmpty()) {
                ui.valueToShowInOverview.text = ui.value.text
            }
        }
        presenter.attachView(this)
        presenter.onViewUpdated(false)
    }

    fun storeBundle(extras: Bundle?) {
        if (extras != null)
            presenter.storeBundle(extras)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        presenter.onSaveInstanceState(outState)
    }

    override fun onStart() {
        super.onStart()
        presenter.attachView(this)
    }

    override fun onResume() {
        super.onResume()
        presenter.refreshExpense()
    }

    override fun onStop() {
        super.onStop()
        presenter.detachView()
    }

    override fun setTitle(string: String) {
        ui.toolbar.title = string
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        presenter.attachView(this)
        when (requestCode) {
            REQUEST_SELECT_CHARGEABLE -> {
                if (resultCode == Activity.RESULT_OK && data != null) {
                    val uuid = data.getStringExtra(ListChargeableActivity.KEY_CHARGEABLE_SELECTED_UUID)
                    val keyChargeableSelectedType = ListChargeableActivity.KEY_CHARGEABLE_SELECTED_TYPE
                    val type = data.getSerializableExtra(keyChargeableSelectedType) as ChargeableType
                    presenter.onChargeableSelected(type, uuid)
                }
            }
            REQUEST_SELECT_BILL -> {
                if (resultCode == Activity.RESULT_OK && data != null)
                    presenter.onBillSelected(data.getStringExtra(ListBillActivity.KEY_BILL_SELECTED_UUID))
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.save, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_save -> presenter.save()
        }
        return super.onOptionsItemSelected(item)
    }

    internal fun onDate() {
        presenter.onDate(this)
    }

    override fun onDateChanged(date: DateTime) {
        ui.date.setText(Transaction.SIMPLE_DATE_FORMAT.format(date.toDate()))
    }

    override fun onChargeableSelected(chargeable: Chargeable) {
        ui.chargeable.setText(chargeable.name)
        ui.payNextMonth.visibility = if (chargeable.canBePaidNextMonth()) View.VISIBLE else View.GONE
    }

    internal fun onChargeable() {
        doAsync {
            val hasChargeable = presenter.hasChargeable()

            uiThread {
                if (!hasChargeable) {
                    val intent = Intent(this@EditExpenseActivity, ListChargeableActivity::class.java)
                    startActivityForResult(intent, REQUEST_SELECT_CHARGEABLE)
                }
            }
        }
    }

    override fun onBillSelected(bill: Bill?) {
        if (bill != null) {
            if (ui.bill.text.toString().isEmpty())
                ui.bill.setText(bill.name)
            if (ui.name.text.toString().isEmpty())
                ui.name.setText(bill.name)
            if (ui.value.text.toString().isEmpty())
                ui.value.setText(bill.amount.toCurrencyFormatted())
            ui.showInOverview.isChecked = false
            ui.showInResume.isChecked = true
        } else {
            ui.bill.setText("")
        }
    }

    internal fun onBill() {
        startActivityForResult(Intent(this, ListBillActivity::class.java), REQUEST_SELECT_BILL)
    }

    override fun fillExpense(expense: Expense): Expense {
        expense.name = ui.name.text.toString()
        var value = 0
        val valueText = ui.value.text.toString().replace("[^\\d]".toRegex(), "")
        if (!StringUtils.isEmpty(valueText))
            value = Integer.parseInt(valueText)

        var valueToShowInOverview = 0
        val valueToShowInOverviewText = ui.valueToShowInOverview.text.toString().replace("[^\\d]".toRegex(), "")
        if (!StringUtils.isEmpty(valueToShowInOverviewText))
            valueToShowInOverview = Integer.parseInt(valueToShowInOverviewText)

        if (ui.repayment.isChecked) {
            value *= -1
            valueToShowInOverview *= -1
        }
        expense.value = value
        expense.valueToShowInOverview = valueToShowInOverview
        expense.chargedNextMonth = ui.payNextMonth.isChecked
        expense.showInOverview(ui.showInOverview.isChecked)
        expense.showInResume(ui.showInResume.isChecked)
        expense.installments = installment
        expense.repetition = repetition
        return expense
    }

    override fun finishView() {
        val i = Intent()
        i.putExtra(KEY_EXPENSE_UUID, presenter.uuid)
        setResult(Activity.RESULT_OK, i)
        finish()
    }

    override fun showError(error: ValidationError) {
        when (error) {
            ValidationError.NAME -> ui.name.error = getString(error.message)
            ValidationError.AMOUNT -> ui.value.error = getString(error.message)
            ValidationError.CHARGEABLE -> ui.chargeable.error = getString(error.message)
            else -> Timber.e("Validation unrecognized, field:" + error)
        }
    }

    private val installment: Int
        get() = Integer.parseInt(ui.installment.text.toString())

    val repetition: Int
        get() = Integer.parseInt(ui.repetition.text.toString())

    override fun showExpense(expense: Expense) {
        ui.name.setText(expense.name)
        ui.value.setText(Math.abs(expense.value).toCurrencyFormatted())
        ui.valueToShowInOverview.setText(Math.abs(expense.valueToShowInOverview).toCurrencyFormatted())
        if (expense.charged) {
            ui.value.setTextColor(ResourcesCompat.getColor(resources, R.color.value_unpaid, null))
            ui.repayment.isEnabled = false
        }
        if (expense.value < 0)
            ui.repayment.isChecked = true

        ui.payNextMonth.isChecked = expense.chargedNextMonth
        ui.showInOverview.isChecked = expense.isShowInOverview
        ui.showInResume.isChecked = expense.isShowInResume
    }

    companion object {
        val KEY_EXPENSE_UUID = ExpensePresenter.KEY_EXPENSE_UUID
        private val REQUEST_SELECT_CHARGEABLE = 1003
        private val REQUEST_SELECT_BILL = 1004
    }
}

private class EditExpenseActivityUi : AnkoComponent<EditExpenseActivity> {
    lateinit var toolbar : TemplateToolbar
    lateinit var name : AppCompatEditText
    lateinit var date : AppCompatEditText
    lateinit var value : AppCompatEditText
    lateinit var valueToShowInOverview : AppCompatEditText
    lateinit var repayment : CheckBox
    lateinit var chargeable : AppCompatEditText
    lateinit var bill : AppCompatEditText
    lateinit var payNextMonth : CheckBox
    lateinit var showInOverview : CheckBox
    lateinit var showInResume : CheckBox
    lateinit var repetition : AppCompatEditText
    lateinit var installment : AppCompatEditText

    override fun createView(ui: AnkoContext<EditExpenseActivity>) = with(ui) {
        verticalLayout {
            toolbar = toolbarTemplate {}

            scrollView {
                padding = resources.getDimensionPixelSize(R.dimen.default_spacing)
                linearLayout {
                    orientation = LinearLayout.VERTICAL
                    textInputLayout {
                        name = appCompatEditText {
                            id = R.id.act_edit_expense_name
                            hint = resources.getString(R.string.name)
                        }
                    }
                    textInputLayout {
                        date = clickableView {
                            id = R.id.act_edit_expense_date
                            hint = resources.getString(R.string.date)
                            onClick { ui.owner.onDate() }
                        }
                    }
                    textInputLayout {
                        value = appCompatEditText {
                            id = R.id.act_edit_expense_value
                            hint = resources.getString(R.string.value)
                            inputType = InputType.TYPE_CLASS_NUMBER
                        }
                    }
                    textInputLayout {
                        valueToShowInOverview = appCompatEditText {
                            id = R.id.act_edit_expense_value_to_show_in_overview
                            hint = resources.getString(R.string.income_to_show_in_overview)
                            inputType = InputType.TYPE_CLASS_NUMBER
                        }
                    }
                    repayment = checkBox {
                        id = R.id.act_edit_expense_repayment
                        text = resources.getString(R.string.repayment)
                    }.lparams {
                        bottomMargin = dip(5)
                    }
                    textInputLayout {
                        chargeable = clickableView {
                            id = R.id.act_edit_expense_chargeable
                            hint = resources.getString(R.string.paid_with)
                            onClick { ui.owner.onChargeable() }
                        }
                    }
                    textInputLayout {
                        bill = clickableView {
                            id = R.id.act_edit_expense_bill
                            hint = resources.getString(R.string.bill)
                            onClick { ui.owner.onBill() }
                        }
                    }
                    payNextMonth = checkBox {
                        id = R.id.act_edit_expense_pay_next_month
                        text = resources.getString(R.string.pay_next_month)
                        visibility = View.GONE
                    }.lparams {
                        bottomMargin = dip(5)
                    }
                    showInOverview = checkBox {
                        id = R.id.act_edit_expense_show_in_overview
                        text = resources.getString(R.string.show_in_overview)
                        isChecked = true
                    }.lparams {
                        bottomMargin = dip(5)
                    }
                    showInResume = checkBox {
                        id = R.id.act_edit_expense_show_in_resume
                        text = resources.getString(R.string.show_in_resume)
                        isChecked = true
                    }.lparams {
                        bottomMargin = dip(5)
                    }
                    textInputLayout {
                        repetition = appCompatEditText {
                            id = R.id.act_edit_expense_repetition
                            hint = resources.getString(R.string.repetition)
                            inputType = InputType.TYPE_CLASS_NUMBER
                            setText(R.string._1)
                        }
                    }
                    textInputLayout {
                        installment = appCompatEditText {
                            id = R.id.act_edit_expense_installment
                            hint = resources.getString(R.string.installments)
                            inputType = InputType.TYPE_CLASS_NUMBER
                            setText(R.string._1)
                        }
                    }
                }.lparams(height = wrapContent, width = matchParent)
            }
        }.applyRecursively(::applyTemplateViewStyles)
    }
}
